src/transformers/models/speecht5/modeling_speecht5.py

```diff
@@ -2587,35 +2595,68 @@ def _generate_speech(
+        )
         # Predict the probability that this is the stop token.
         prob = torch.sigmoid(model.speech_decoder_postnet.prob_out(last_decoder_output))
 
-        # Finished when stop token or maximum length is reached.
-        if idx >= minlen and (int(sum(prob >= threshold)) > 0 or idx >= maxlen):
-            spectrogram = torch.cat(spectrogram, dim=0).unsqueeze(0)
-            spectrogram = model.speech_decoder_postnet.postnet(spectrogram)
-            spectrogram = spectrogram.squeeze(0)
-            break
-
-    if vocoder is not None:
-        outputs = vocoder(spectrogram)
+        if idx < minlen:
+            continue
+        else:
+            spectrograms = torch.stack(spectrogram)
+            spectrograms = spectrograms.transpose(0, 1).flatten(1, 2)
+            spectrograms = model.speech_decoder_postnet.postnet(spectrograms)
+
+            meet_thresholds = torch.sum(prob, dim=-1) >= threshold
+            # If the generation loops is less than maximum length times, check the ones in the batch that have met
+            # the prob threshold. Otherwise, assume all have met thresholds and fill other spectrograms for the batch.
+            meet_indexes = torch.where(meet_thresholds)[0] if idx < maxlen else torch.tensor(list(range(len(prob))))
+            meet_indexes = meet_indexes.tolist()
+            for meet_index in meet_indexes:
+                if meet_index in result_spectrogram:
+                    continue
+                result_spectrogram[meet_index] = spectrograms[meet_index]
+
+            if len(result_spectrogram) >= bsz:
+                break
+
+    spectrograms = [result_spectrogram[i] for i in range(len(result_spectrogram))]
+    if bsz == 1:
+        spectrogram = spectrograms[0]
+        if vocoder is not None:
+            outputs = vocoder(spectrogram)
+        else:
+            outputs = spectrogram
+        if output_cross_attentions:
+            cross_attentions = torch.cat(cross_attentions, dim=2)
+            outputs = (outputs, cross_attentions)
     else:
-        outputs = spectrogram
-
-    if output_cross_attentions:
-        cross_attentions = torch.cat(cross_attentions, dim=2)
-        outputs = (outputs, cross_attentions)
+        # batched return values should also include the spectrogram/waveform lengths
+        if vocoder is None:
+            spectrogram_lengths = []
+            for i in range(bsz):
+                spectrogram_lengths.append(spectrograms[i].size(0))
+            spectrograms = torch.cat(spectrograms, dim=0)
+            outputs = (spectrograms, spectrogram_lengths)
+        else:
+            waveform_lengths = []
+            waveforms = []
+            for i in range(bsz):
+                waveform = vocoder(spectrograms[i])
+                waveforms.append(waveform)
+                waveform_lengths.append(waveform.size(0))
+            waveforms = torch.cat(waveforms, dim=0)
+            outputs = (waveforms, waveform_lengths)
```

Comment:
This would be great to also have an option to return `cross_attention` when batching. At the moment `output_cross_attentions` do nothing if a batch is passed into inputs.

Reply:
Sure, added in latest commit.